File: Calculus/integrals.py
import numpy as np


def simpson(f, a, b, n):
    """Approximates the definite integral of f from a to b by
    the composite Simpson's rule, using n subintervals.
    From http://en.wikipedia.org/wiki/Simpson's_rule
    """
    h = (b - a) / n
    i = np.arange(0,n)
    
    s = f(a) + f(b) 
    s += 4 * np.sum( f( a + i[1::2] * h ) )
    s += 2 * np.sum( f( a + i[2:-1:2] * h ) )
    
    return s * h / 3

def trapezoid(f, a, b, n):
    """Approximates the definite integral of f from a to b by
    the composite trapezoidal rule, using n subintervals.
    From http://en.wikipedia.org/wiki/Trapezoidal_rule
    """
    h = (b - a) / n
    s = f(a) + f(b)
    i = np.arange(0,n)
    s += 2 * np.sum( f(a + i[1:] * h) )
    return s * h / 2


def adaptive_trapezoid(f, a, b, acc, output=False):
    """
    Uses the adaptive trapezoidal method to compute the definite integral
    of f from a to b to desired accuracy acc. 
    """
    old_s = np.inf
    h = b - a
    n = 1
    s = (f(a) + f(b)) * 0.5
    if output == True : 
        print ("N = " + str(n+1) + ",  Integral = " + str( h*s ))
    while abs(h * (old_s - s*0.5)) > acc :
        old_s = s
        for i in np.arange(n) :
            s += f(a + (i + 0.5) * h)
        n *= 2.
        h *= 0.5
        if output == True :
            print ("N = " + str(n) + ",  Integral = " + str( h*s ))
    return h * s

File: Calculus/integrals.cc
#include <vector>
#include <iostream>
#include <math.h>
#include "integrals.h"

using namespace std;


int main()
{
  unsigned n1=1,n2=0;
  cout << "Trapezoid quadrature of sin(x) [0, pi/2]" << endl;
  while ( n1 % 2 != 0 ) {
    cout << "Enter number of intervals desired for trapezoidal rule (must be even)" << endl;  
    cin >> n1;
  }
  double a = 0;
  double b = 2 * atan(1.0);
  double ans1 = trapezoid<double>(sin, a, b, n1);
  cout << "Trapezoidal rule : " << ans1 << endl;

  double ans2 = simpson<double>(sin, a, b, n1);
  cout << "Trapezoidal rule : " << ans2 << endl;

  double ans3 = adaptive_trapezoid<double>( sin, a, b, 0.0001);
  cout << "Adaptive trapezoidal rule : " << ans3 << endl;
}
